### ISSUE TITLE
botsort.yaml and with with_reid 

### ISSUE THREAD

**Issue by KruemelmonsterK** [ultralytics/ultralytics#3818]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey guys,

i want to do tracking of mice in real-time with re-identification.

There are BoT-SORT and BoT-SORT-ReID, but you only offer botsort.yml. Do I use  BoT-SORT-ReID, when I set with with_reid: True?

When a mouse leaves the Field of View and enters again, does it get the same ID assigned or a different ID?

Besides I wonder if i can only fine tune yolov8 or can I fine tune BoT-SORT or ByteTrack as well somehow?

**Comment by KruemelmonsterK** [comment]
Addendum: So after e.g. occlusion I need the possibility to reidentify the mouse and assign it the same id as before. Reidentification even needs to be possible when multipble mice where not visiblae for 5 minutes and then appear again. I guess this involves learning some features of the mouse to reidentifiy it. Is this possible with yolov8 with tracking enabled or do I need to add my own classificator to differentiat the mice e.g. by their face?

**Comment by glenn-jocher** [comment]
@KruemelmonsterK hello,

Thank you for your question and the additional information.

You're absolutely correct in understanding the need for feature learning to ensure robust re-identification under occlusions or when mice disappear and reappear in the frame.

In standard tracking algorithms like BoT-SORT, identity switches can occur when objects re-enter the scene after a period of absence. To overcome this, you would indeed require a re-identification mechanism that studies the distinct, learned features of tracked objects.

With YOLOv8's tracking features, the underlying object detector is responsible for identifying and classifying objects in the camera's field of view (e.g., mice), but it does not inherently support reidentification after object occlusions or disappearances.

As you've pointed out, integrating a separate classifier might be the solution you need. This detector can be trained to recognize individual mice based on unique features (like faces, coloring, markings, etc.). When an identified mouse disappears and re-enters the field of view, this classifier would allow the tracking system to maintain continuity by assigning the previously used ID to this mouse.

The use of feature embedding and a variety of deep learning techniques are common approaches for these kinds of "person re-identification" challenges, even though they'd be used in your case for mice. These systems work by learning a deep representation of each object, which can be used to differentiate between objects, even in the presence of occlusions or temporary disappearances.

So, to summarize, while YOLOv8 provides excellent capabilities for object detection and tracking, for your specific case where reidentification is required after occlusions or periods of absence, you would likely need to integrate additional capabilities in the form of a distinct feature classifier.

Again, please don't hesitate to reply if you need more clarification or have additional questions!

**Comment by wolfgang1021** [comment]
@glenn-jocher @KruemelmonsterK 
hello guys，
# ReID model related thresh (not supported yet)
proximity_thresh: 0.5
appearance_thresh: 0.25
with_reid: False

(not supported yet) Does it mean that when I set "with_reid" to "True", I still get Bot-Sort, not Bot-Sort-Reid. Thanks for your help.

**Comment by AmineMekki01** [comment]
@KruemelmonsterK and @glenn-jocher 
Were you able to solve that issue?
I'm working on something similar. I want to detect, identify, re-identify, and track two rats in the same cage, while also incorporating a thermal camera to measure their temperature. Accurate re-identification is critical to ensure that temperature profiles aren't mixed up, especially in cases of occlusions or temporary disappearances. what are your thoughts on this?

**Comment by glenn-jocher** [comment]
@AmineMekki01 yOLOv8's tracking combined with the `with_reid=True` option in BoT-SORT can help with re-identification by utilizing appearance features, but for highly accurate re-identification over occlusions or extended absences, you might need a custom feature extractor. Training a classifier to uniquely identify and re-identify the two rats would improve reliability, especially when working with thermal imagery. Consider merging detection/tracking outputs with thermal data for unique signatures. For additional customization, you can modify or extend the BoT-SORT pipeline. Let me know if you need help setting this up.

**Comment by ibrahim907** [comment]
Hi guys,
How  botsort can modified and extended to support simple re-identification (in  my case trucks).
After trucks Occulsion And What features  extraction models you suggest  . I'm  a beginner in this this field 
Sorry for any grammar mistakes.

**Comment by glenn-jocher** [comment]
Hi, BoT-SORT can be extended for re-identification by incorporating a feature extraction model, such as ResNet or MobileNet, to generate embeddings for each truck. These embeddings can then be used to match identities post-occlusion. You can modify the `botsort.py` to include a ReID module. For beginners, I recommend starting with pretrained models for feature extraction to simplify the process. Let me know if you need further clarification!

**Comment by ibrahim907** [comment]
@glenn-jocher
Thank for your answer,
I need further clarification
I using bot sort From boxmot package which has feature extracter .

**Comment by glenn-jocher** [comment]
@ibrahim907 thank you for clarifying. If you're using BoT-SORT from the boxmot package with its built-in feature extractor, you'll need to ensure the `with_reid=True` option is enabled and the feature extractor is properly trained or fine-tuned to handle your specific truck re-identification use case. For further customization or issues specific to boxmot, I recommend consulting their documentation or support channels. Let me know if you need help related to Ultralytics' implementation of BoT-SORT!

**Comment by chrizwill** [comment]
@glenn-jocher could you explain further how exactly I would go about incorporating a feature extraction model and modifying the botsort.py file to include a ReID module? Thanks in advance!

**Comment by glenn-jocher** [comment]
@chrizwill to enhance BoT-SORT with ReID in Ultralytics, enable the `with_reid` parameter in your tracker YAML config and ensure detections include appearance features. The existing `BOTrack` class in [BOTSORT](https://docs.ultralytics.com/reference/trackers/bot_sort/#ultralytics.trackers.bot_sort.BOTSORT) already supports feature storage in its `update_features` method. For implementation details, see our [track mode documentation](https://docs.ultralytics.com/modes/track/). Let me know if you need specifics about Ultralytics' built-in tracking implementation!

**Comment by chrizwill** [comment]
@glenn-jocher so I get the "setting with_reid: True" part, what do you mean by "ensure detections include appearance features"? If you could explain what I need to do beyond changing the with_reid argument in the yaml file, that would be helpful. Would I need to change anything in the bot_sort.py file and how would I go about integrating a separate reid model (would that even be necessary?)

**Comment by chrizwill** [comment]
@glenn-jocher When looking at the bot_sort.py code I see that ReID is not actually implemented right? The code explicitly states "Haven't supported BoT-SORT(reid) yet" and the Re-ID encoder is initialized as None. While there are some components designed to support Re-ID in the future, such as methods for handling features and calculating embedding distances, the actual Re-ID functionality appears to be conditional. What do I need to do to get some actual ReID functionality?

**Comment by glenn-jocher** [comment]
You're correct that ReID isn't fully implemented in Ultralytics' BOTSORT yet. To enable actual ReID functionality, you would need to either:  
1. Use an external implementation like [BoT-SORT with ReID from boxmot](https://github.com/mikel-brostrom/boxmot) which has full ReID support  
2. Modify `bot_sort.py` to integrate a ReID model by implementing the encoder logic (currently set to `None` when `with_reid=True`)  
Our implementation focuses on the core tracking components, but we're tracking ReID integration requests for future updates. For now, boxmot's implementation may better suit ReID needs.

**Comment by glenn-jocher** [comment]
@chrizwill Great news! As UltralyticsAssistant mentioned, proper ReID support has just been merged into Ultralytics in PR #20192. This implementation adds the functionality you're looking for.

Looking at the reference docs for `bot_sort.py`, the BOTSORT class now properly supports ReID through its appearance-based matching capabilities. The implementation uses feature embeddings to track objects across frames, even when they temporarily disappear.

To use this new functionality:

1. Upgrade to the latest version as suggested (`pip install -U ultralytics>=8.3.114`)
2. Enable ReID in your tracker config with `with_reid: True`
3. You can specify which model to use for feature extraction with the `model` parameter

The ReID module works by extracting embeddings for each detected object, which are then used for appearance-based matching alongside the standard IoU-based tracking. This should help maintain identity even after occlusions or when objects leave and re-enter the frame.

Let me know if you encounter any issues with the new implementation!

**Comment by dkarvan8** [comment]
Hello, I'm currently using YOLO11n with BoT-SORT and the new ReID functionality (v8.3.114+). I understand the model: auto option, but I'm interested in using a custom ReID feature extraction model. For instance, if I've trained my own torchreid model and saved its weights as a .pt file or any pretrained model, can I simply put this .pt file in a specific directory (e.g., ultralytics/models or a custom path) and then reference its filename in the model parameter of my botsort.yaml (e.g., model: my_custom_reid_model.pt)? Or is there a different process for integrating custom ReID models



Regarding customizing the tracker configuration, what is the recommended best practice for modifying parameters or specifying a custom ReID feature extraction model?

Should I:

Directly edit the botsort.yaml file located within the Ultralytics installation directory (e.g., ultralytics/cfg/trackers/botsort.yaml)?

Create a new, separate YAML file (e.g., my_custom_tracker_config.yaml) in my project directory, copy the default botsort.yaml content into it, make my desired changes there, and then pass the path to this new file via the tracker= argument (e.g., model.track(..., tracker='path/to/my_custom_tracker_config.yaml'))?

**Comment by glenn-jocher** [comment]
@dkarvan8 for custom ReID models, the current implementation specifically uses YOLO classification models rather than arbitrary torchreid models. You can specify models like `yolo11n-cls.pt`, `yolo11s-cls.pt`, etc., or use `model: auto` for native features. Custom torchreid models aren't directly supported - the ReID module expects YOLO classification model formats.

For tracker configuration, the recommended approach is option 2: create a separate custom YAML file in your project directory, copy the default botsort.yaml content, modify it there, and reference it with `tracker='path/to/my_custom_tracker_config.yaml'`. This keeps your customizations separate from the installation and makes your project more portable.

**Comment by dkarvan8** [comment]
thank you for taking the time to answer!!
how do I check the accuracy of the models?

**Comment by glenn-jocher** [comment]
@dkarvan8 you can evaluate model accuracy using YOLO's validation mode with `model.val()` or `yolo val model=yolo11n.pt data=your_dataset.yaml` for detection accuracy. For tracking performance, you'll need ground truth annotations with object IDs to calculate metrics like MOTA, MOTP, and IDF1. The [benchmark mode](https://docs.ultralytics.com/modes/benchmark/) can help evaluate inference speed across different formats. For ReID specifically, you'd typically measure embedding similarity accuracy between same/different object pairs if you have labeled data.

**Comment by Hemilibeatriz** [comment]
Which REID model is used by default in Ultralytics BOTSORT?

**Comment by dkarvan8** [comment]
> Which REID model is used by default in Ultralytics BOTSORT?

yolo

**Comment by Hemilibeatriz** [comment]
> > Which REID model is used by default in Ultralytics BOTSORT?
> 
> yolo

thanks @dkarvan8, do you know if i can implement an transreid model instead?

**Comment by dkarvan8** [comment]
> > > Which REID model is used by default in Ultralytics BOTSORT?
> > 
> > 
> > yolo
> 
> thanks [@dkarvan8](https://github.com/dkarvan8), do you know if i can implement an transreid model instead?

hey give it a try. I tried integrating osnet model, but there were latency and accuracy issues.

**Comment by Hemilibeatriz** [comment]
> > > > Which REID model is used by default in Ultralytics BOTSORT?
> > > 
> > > 
> > > yolo
> > 
> > 
> > thanks [@dkarvan8](https://github.com/dkarvan8), do you know if i can implement an transreid model instead?
> 
> hey give it a try. I tried integrating osnet model, but there were latency and accuracy issues.

I understand @dkarvan8 i will try

**Comment by pranavsabbu1** [comment]
@glenn-jocher Hello, my goal is to create an object tracking algorithm that can handle occlusion and group settings. Currently I am using botsort implementation by ultralytics. This is how I set my .yaml file. Is this the proper way to use re-id with bot sort? 

tracker_type: botsort # Type of Track

track_high_thresh: 0.6 # confidence threshold for the initial assocation of a detection with a track
track_low_thresh: 0.1 # second association attempt if first one fails
new_track_thresh: 0.85 # threshold required for a detection to initiate a new track

track_buffer: 2000 # threshold for spatial proximity when associating tracks and detections

match_thresh: 0.9 
proximity_thresh: 0.3 # threshold for spatial proximity when associating tracks and detections
appearance_thresh: 0.75 # threshold for appearance similarity (based on ReID)

with_reid: True
model: auto

gmc_method: sparseOptFlow # instance of the global motion compensation algorithm ``

fuse_score: True

**Comment by glenn-jocher** [comment]
Thanks—yes, that YAML enables ReID in BoT‑SORT; `with_reid: True` with `model: auto` uses YOLO-native features (or set a YOLO classification model like `yolo11n-cls.pt`), and ReID is fused with IoU inside BOTSORT’s distance calculation; see the BOTSORT class reference for details. [BOTSORT class reference](https://docs.ultralytics.com/reference/trackers/bot_sort/). ([docs.ultralytics.com](https://docs.ultralytics.com/reference/trackers/bot_sort/))

A couple of tweaks: `track_buffer` is the number of frames to keep lost tracks alive (not a spatial threshold), `new_track_thresh: 0.85` may block new IDs if your detector confidences are lower, and `match_thresh: 0.9` is very lenient and can increase ID switches—start from the defaults in the default botsort.yaml and adjust gradually. [Default botsort.yaml](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/trackers/botsort.yaml). ([github.com](https://github.com/ultralytics/ultralytics/issues/20498))

If you’re feeding frames one-by-one, remember to persist IDs across calls:
```python
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
results = model.track("video.mp4", tracker="my_botsort.yaml", persist=True)
```
